File: src/ee/admob/private/AdMobRewardedAd.swift
//
//  AdMobRewardedAd.swift
//  ee-x-366f64ba
//
//  Created by eps on 6/24/20.
//

import Foundation
import GoogleMobileAds

internal class AdMobRewardedAd: NSObject, GADRewardedAdDelegate {
    private let _logger = Logger("\(AdMobRewardedAd.self)")
    private let _bridge: IMessageBridge
    private let _adId: String
    private let _messageHelper: MessageHelper
    private var _isLoaded = false
    private var _rewarded = false
    private var _ad: GADRewardedAd?
    
    init(_ bridge: IMessageBridge,
         _ adId: String) {
        _bridge = bridge
        _adId = adId
        _messageHelper = MessageHelper("AdMobRewardedAd", _adId)
        super.init()
        registerHandlers()
        createInternalAd()
    }
    
    func destroy() {
        deregisterHandlers()
        destroyInternalAd()
    }
    
    func registerHandlers() {
        _bridge.registerHandler(_messageHelper.createInternalAd) { _ in
            self.createInternalAd()
            return ""
        }
        _bridge.registerHandler(_messageHelper.destroyInternalAd) { _ in
            self.destroyInternalAd()
            return ""
        }
        _bridge.registerHandler(_messageHelper.isLoaded) { _ in
            Utils.toString(self.isLoaded)
        }
        _bridge.registerHandler(_messageHelper.load) { _ in
            self.load()
            return ""
        }
        _bridge.registerHandler(_messageHelper.show) { _ in
            self.show()
            return ""
        }
    }
    
    func deregisterHandlers() {
        _bridge.deregisterHandler(_messageHelper.createInternalAd)
        _bridge.deregisterHandler(_messageHelper.destroyInternalAd)
        _bridge.deregisterHandler(_messageHelper.isLoaded)
        _bridge.deregisterHandler(_messageHelper.load)
        _bridge.deregisterHandler(_messageHelper.show)
    }
    
    func createInternalAd() {
        Thread.runOnMainThread {
            if self._ad != nil {
                return
            }
            self._ad = GADRewardedAd(adUnitID: self._adId)
        }
    }
    
    func destroyInternalAd() {
        Thread.runOnMainThread {
            if self._ad == nil {
                return
            }
            self._ad = nil
        }
    }
    
    var isLoaded: Bool {
        return _isLoaded
    }
    
    func load() {
        Thread.runOnMainThread {
            guard let ad = self._ad else {
                assert(false, "Ad is not initialized")
            }
            ad.load(GADRequest(), completionHandler: { error in
                if let error = error {
                    self._bridge.callCpp(self._messageHelper.onFailedToLoad,
                                         error.localizedDescription)
                } else {
                    self._bridge.callCpp(self._messageHelper.onLoaded)
                }
            })
        }
    }
    
    func show() {
        Thread.runOnMainThread {
            guard
                let ad = self._ad,
                let rootView = Utils.getCurrentRootViewController()
            else {
                assert(false, "Ad is not initialized")
            }
            ad.present(fromRootViewController: rootView, delegate: self)
        }
    }
    
    func rewardedAdDidPresent(_ rewardedAd: GADRewardedAd) {
        _logger.debug("\(#function)")
    }
    
    func rewardedAd(_ rewardedAd: GADRewardedAd, didFailToPresentWithError error: Error) {
        _logger.debug("\(#function): \(error.localizedDescription)")
        _bridge.callCpp(_messageHelper.onFailedToShow, error.localizedDescription)
    }
    
    func rewardedAd(_ rewardedAd: GADRewardedAd, userDidEarn reward: GADAdReward) {
        _logger.debug("\(#function)")
        _rewarded = true
    }
    
    func rewardedAdDidDismiss(_ rewardedAd: GADRewardedAd) {
        _logger.debug("\(#function)")
        _isLoaded = false
        _bridge.callCpp(_messageHelper.onClosed, Utils.toString(_rewarded))
    }
}

File: src/android/admob/src/main/java/com/ee/internal/AdMobRewardedAd.kt
package com.ee.internal

import android.app.Activity
import androidx.annotation.AnyThread
import com.ee.ILogger
import com.ee.IMessageBridge
import com.ee.Thread
import com.ee.Utils
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.rewarded.RewardItem
import com.google.android.gms.ads.rewarded.RewardedAd
import com.google.android.gms.ads.rewarded.RewardedAdCallback
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback
import com.google.common.truth.Truth.assertThat
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Created by KietLe on 12/03/19.
 */
internal class AdMobRewardedAd(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger,
    private var _activity: Activity?,
    private val _adId: String) : RewardedAdCallback() {
    companion object {
        private val kTag = AdMobRewardedAd::class.java.name
    }

    private val _messageHelper = MessageHelper("AdMobRewardedAd", _adId)
    private val _isLoaded = AtomicBoolean(false)
    private var _rewarded = false
    private var _ad: RewardedAd? = null

    init {
        _logger.info("$kTag: constructor: adId = $_adId")
        registerHandlers()
        createInternalAd()
    }

    fun onCreate(activity: Activity) {
        _activity = activity
    }

    fun onDestroy(activity: Activity) {
        assertThat(_activity).isEqualTo(activity)
        _activity = null
    }

    @AnyThread
    fun destroy() {
        _logger.info("$kTag: destroy: adId = $_adId")
        deregisterHandlers()
        destroyInternalAd()
    }

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerHandler(_messageHelper.isLoaded) {
            Utils.toString(isLoaded)
        }
        _bridge.registerHandler(_messageHelper.load) {
            load()
            ""
        }
        _bridge.registerHandler(_messageHelper.show) {
            show()
            ""
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(_messageHelper.isLoaded)
        _bridge.deregisterHandler(_messageHelper.load)
        _bridge.deregisterHandler(_messageHelper.show)
    }

    @AnyThread
    private fun createInternalAd() {
        Thread.runOnMainThread {
            if (_ad != null) {
                return@runOnMainThread
            }
            // https://github.com/googleads/googleads-mobile-android-mediation/issues/113
            // https://developers.google.com/admob/android/mediate#initialize_your_ad_object_with_an_activity_instance
            _ad = RewardedAd(_activity, _adId)
        }
    }

    @AnyThread
    private fun destroyInternalAd() {
        Thread.runOnMainThread {
            if (_ad == null) {
                return@runOnMainThread
            }
            _ad = null
        }
    }

    private val isLoaded: Boolean
        @AnyThread get() = _isLoaded.get()

    @AnyThread
    private fun load() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::load.name}")
            val ad = _ad ?: throw IllegalArgumentException("Ad is not initialized")
            val callback = object : RewardedAdLoadCallback() {
                override fun onRewardedAdLoaded() {
                    Thread.runOnMainThread {
                        _logger.debug("${kTag}: ${this::onRewardedAdLoaded.name}")
                        _isLoaded.set(true)
                        _bridge.callCpp(_messageHelper.onLoaded)
                    }
                }

                override fun onRewardedAdFailedToLoad(error: LoadAdError?) {
                    Thread.runOnMainThread {
                        _logger.debug("${kTag}: onRewardedAdFailedToLoad: message = ${error?.message ?: ""} response = ${error?.responseInfo ?: ""}")
                        destroyInternalAd()
                        createInternalAd()
                        _bridge.callCpp(_messageHelper.onFailedToLoad, error?.message ?: "")
                    }
                }
            }
            ad.loadAd(AdRequest.Builder().build(), callback)
        }
    }

    @AnyThread
    private fun show() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::show.name}")
            val ad = _ad ?: throw IllegalArgumentException("Ad is not initialized")
            _rewarded = false
            ad.show(_activity, this)
        }
    }

    override fun onRewardedAdFailedToShow(error: AdError?) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: onRewardedAdFailedToShow: message = ${error?.message ?: ""}")
            destroyInternalAd()
            createInternalAd()
            _bridge.callCpp(_messageHelper.onFailedToShow, error?.message ?: "")
        }
    }

    override fun onRewardedAdOpened() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onRewardedAdOpened.name}")
            _isLoaded.set(false)
        }
    }

    override fun onUserEarnedReward(reward: RewardItem) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onUserEarnedReward.name}")
            _rewarded = true
        }
    }

    override fun onRewardedAdClosed() {
        Thread.runOnMainThread {
            _logger.info("$kTag: ${this::onRewardedAdClosed.name}")
            destroyInternalAd()
            createInternalAd()
            _bridge.callCpp(_messageHelper.onClosed, Utils.toString(_rewarded))
        }
    }
}